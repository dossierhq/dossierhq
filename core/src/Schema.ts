import type { EntityReference, Result, ErrorType, Location, RichText, ValueItem } from '.';
import { notOk, ok } from '.';

export interface EntityTypeSpecification {
  name: string;
  fields: FieldSpecification[];
}

export interface ValueTypeSpecification {
  name: string;
  fields: FieldSpecification[];
}

export enum FieldType {
  EntityType = 'EntityType',
  Location = 'Location',
  RichText = 'RichText',
  String = 'String',
  ValueType = 'ValueType',
}

export enum RichTextBlockType {
  entity = 'entity',
  valueItem = 'valueItem',
  paragraph = 'paragraph',
}

export interface FieldSpecification {
  name: string;
  type: FieldType;
  list?: boolean;
  isName?: boolean;
  /** Applicable when type is EntityType or RichText */
  entityTypes?: string[];
  /** Applicable when type is ValueType or RichText */
  valueTypes?: string[];
  /** Applicable when type is RichText. All block types are enabled if none are specified. The type
   * can either be a standard RichTextBlockType or any type that's enabled.
   * If inlineTypes isn't specified all inline types are enabled. If inlineTypes is an empty array,
   * no inline types are enabled.
   */
  richTextBlocks?: { type: RichTextBlockType | string; inlineTypes?: string[] }[];
}

export interface FieldValueTypeMap {
  [FieldType.EntityType]: EntityReference;
  [FieldType.Location]: Location;
  [FieldType.RichText]: RichText;
  [FieldType.String]: string;
  [FieldType.ValueType]: ValueItem;
}

export interface SchemaSpecification {
  entityTypes: EntityTypeSpecification[];
  valueTypes: ValueTypeSpecification[];
}

export class Schema {
  readonly spec: SchemaSpecification;

  constructor(spec: SchemaSpecification) {
    this.spec = spec;
  }

  validate(): Result<void, ErrorType.BadRequest> {
    const usedNames = new Set();
    for (const typeSpec of [...this.spec.entityTypes, ...this.spec.valueTypes]) {
      if (usedNames.has(typeSpec.name)) {
        return notOk.BadRequest(`${typeSpec.name}: Duplicate type name`);
      }
      usedNames.add(typeSpec.name);

      for (const fieldSpec of typeSpec.fields) {
        if (!(fieldSpec.type in FieldType)) {
          return notOk.BadRequest(
            `${typeSpec.name}.${fieldSpec.name}: Specified type ${fieldSpec.type} doesn’t exist`
          );
        }

        if (fieldSpec.entityTypes && fieldSpec.entityTypes.length > 0) {
          if (fieldSpec.type !== FieldType.EntityType && fieldSpec.type !== FieldType.RichText) {
            return notOk.BadRequest(
              `${typeSpec.name}.${fieldSpec.name}: Field with type ${fieldSpec.type} shouldn’t specify entityTypes`
            );
          }
          for (const referencedTypeName of fieldSpec.entityTypes) {
            if (this.spec.entityTypes.findIndex((x) => x.name === referencedTypeName) < 0) {
              return notOk.BadRequest(
                `${typeSpec.name}.${fieldSpec.name}: Referenced entity type in entityTypes ${referencedTypeName} doesn’t exist`
              );
            }
          }
        }

        if (fieldSpec.valueTypes && fieldSpec.valueTypes.length > 0) {
          if (fieldSpec.type !== FieldType.ValueType && fieldSpec.type !== FieldType.RichText) {
            return notOk.BadRequest(
              `${typeSpec.name}.${fieldSpec.name}: Field with type ${fieldSpec.type} shouldn’t specify valueTypes`
            );
          }
          for (const referencedTypeName of fieldSpec.valueTypes) {
            if (this.spec.valueTypes.findIndex((x) => x.name === referencedTypeName) < 0) {
              return notOk.BadRequest(
                `${typeSpec.name}.${fieldSpec.name}: Value type in valueTypes ${referencedTypeName} doesn’t exist`
              );
            }
          }
        }

        if (fieldSpec.richTextBlocks && fieldSpec.richTextBlocks.length > 0) {
          if (fieldSpec.type !== FieldType.RichText) {
            return notOk.BadRequest(
              `${typeSpec.name}.${fieldSpec.name}: Field with type ${fieldSpec.type} shouldn’t specify richTextBlocks`
            );
          }

          const paragraph = fieldSpec.richTextBlocks.find(
            (x) => x.type === RichTextBlockType.paragraph
          );
          if (!paragraph) {
            return notOk.BadRequest(
              `${typeSpec.name}.${fieldSpec.name}: richTextBlocks must include paragraph`
            );
          }

          const usedRichTextBlockTypes = new Set();
          for (const richTextBlock of fieldSpec.richTextBlocks) {
            if (usedRichTextBlockTypes.has(richTextBlock.type)) {
              return notOk.BadRequest(
                `${typeSpec.name}.${fieldSpec.name}: richTextBlocks with type ${richTextBlock.type} is duplicated`
              );
            }
            usedRichTextBlockTypes.add(richTextBlock.type);

            if (
              (richTextBlock.type === RichTextBlockType.entity ||
                richTextBlock.type === RichTextBlockType.valueItem) &&
              richTextBlock.inlineTypes &&
              richTextBlock.inlineTypes.length > 0
            ) {
              return notOk.BadRequest(
                `${typeSpec.name}.${fieldSpec.name}: richTextBlocks with type ${richTextBlock.type} shouldn’t specify inlineTypes`
              );
            }
          }
        }
      }
    }

    return ok(undefined);
  }

  getEntityTypeCount(): number {
    return this.spec.entityTypes.length;
  }

  getEntityTypeSpecification(type: string): EntityTypeSpecification | null {
    return this.spec.entityTypes.find((x) => x.name === type) ?? null;
  }

  getEntityFieldSpecification(
    entitySpec: EntityTypeSpecification,
    fieldName: string
  ): FieldSpecification | null {
    return entitySpec.fields.find((x) => x.name === fieldName) ?? null;
  }

  getValueTypeCount(): number {
    return this.spec.valueTypes.length;
  }

  getValueTypeSpecification(type: string): ValueTypeSpecification | null {
    return this.spec.valueTypes.find((x) => x.name === type) ?? null;
  }

  getValueFieldSpecification(
    valueSpec: ValueTypeSpecification,
    fieldName: string
  ): FieldSpecification | null {
    return valueSpec.fields.find((x) => x.name === fieldName) ?? null;
  }
}
