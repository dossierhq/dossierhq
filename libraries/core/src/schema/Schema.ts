import { ok, type ErrorType, type Result } from '../ErrorResult.js';
import { BaseSchema } from './BaseSchema.js';
import { PublishedSchema } from './PublishedSchema.js';
import type {
  AdminSchemaMigrationAction,
  AdminSchemaSpecification,
  AdminSchemaSpecificationUpdate,
  AdminSchemaSpecificationWithMigrations,
} from './SchemaSpecification.js';
import { schemaAdminToPublished } from './schemaAdminToPublished.js';
import { schemaUpdate } from './schemaUpdate.js';
import { schemaValidateAdmin } from './schemaValidateAdmin.js';

export class Schema<
  TSpec extends
    | AdminSchemaSpecification
    | AdminSchemaSpecificationWithMigrations = AdminSchemaSpecification,
> extends BaseSchema<TSpec> {
  private cachedPublishedSchema: PublishedSchema | null = null;

  static createAndValidate(
    update: AdminSchemaSpecificationUpdate,
  ): Result<Schema, typeof ErrorType.BadRequest> {
    const result = SchemaWithMigrations.createAndValidate(update);
    if (!result.isOk()) return result;

    const { migrations, ...specWithoutMigrations } = result.value.spec;

    return ok(new Schema(specWithoutMigrations));
  }

  constructor(spec: TSpec) {
    super(spec);
  }

  validate(): Result<void, typeof ErrorType.BadRequest> {
    return schemaValidateAdmin(this);
  }

  toPublishedSchema(): PublishedSchema {
    if (this.cachedPublishedSchema) {
      return this.cachedPublishedSchema;
    }

    const publishedSpec = schemaAdminToPublished(this);

    this.cachedPublishedSchema = new PublishedSchema(publishedSpec);
    return this.cachedPublishedSchema;
  }
}

export class SchemaWithMigrations extends Schema<AdminSchemaSpecificationWithMigrations> {
  static override createAndValidate(
    update: AdminSchemaSpecificationUpdate,
  ): Result<SchemaWithMigrations, typeof ErrorType.BadRequest> {
    const emptySpec: AdminSchemaSpecificationWithMigrations = {
      schemaKind: 'full',
      version: 0,
      entityTypes: [],
      componentTypes: [],
      patterns: [],
      indexes: [],
      migrations: [],
    };
    const empty = new SchemaWithMigrations(emptySpec);
    return empty.updateAndValidate(update);
  }

  updateAndValidate(
    update: AdminSchemaSpecificationUpdate,
  ): Result<SchemaWithMigrations, typeof ErrorType.BadRequest> {
    // Update
    const updatedResult = schemaUpdate(this.spec, update);
    if (updatedResult.isError()) return updatedResult;
    const updatedSpec = updatedResult.value;

    if (updatedSpec === this.spec) {
      return ok(this); // no change
    }

    // Validate
    const updatedSchema = new SchemaWithMigrations(updatedSpec);
    const validateResult = updatedSchema.validate();
    if (validateResult.isError()) return validateResult;

    return ok(updatedSchema);
  }

  collectMigrationActionsSinceVersion(oldSchemaVersion: number): AdminSchemaMigrationAction[] {
    const migrationsToConsider = this.spec.migrations.filter((it) => it.version > oldSchemaVersion);
    if (migrationsToConsider.length === 0) {
      return [];
    }

    migrationsToConsider.sort((a, b) => a.version - b.version);

    const actions = [];
    for (const migration of migrationsToConsider) {
      actions.push(...migration.actions);
    }
    return actions;
  }
}
