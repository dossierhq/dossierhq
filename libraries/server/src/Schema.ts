import type {
  AdminSchemaSpecification,
  AdminSchemaSpecificationUpdate,
  ErrorType,
  PromiseResult,
  SchemaSpecificationUpdatePayload,
} from '@jonasb/datadata-core';
import { AdminSchema, ok } from '@jonasb/datadata-core';
import type { DatabaseAdapter, TransactionContext } from '@jonasb/datadata-database-adapter';

export async function getSchemaSpecification(
  databaseAdapter: DatabaseAdapter,
  context: TransactionContext
): PromiseResult<AdminSchemaSpecification, ErrorType.Generic> {
  const { logger } = context;
  logger.info('Loading schema');
  const result = await databaseAdapter.schemaGetSpecification(context);
  if (result.isError()) {
    return result;
  }

  const specification = result.value;
  if (!specification) {
    logger.info('No schema set, defaulting to empty');
    return ok({ entityTypes: [], valueTypes: [] });
  }
  logger.info(
    'Loaded schema with %d entity types and  %d value types',
    specification.entityTypes.length,
    specification.valueTypes.length
  );
  return ok(specification);
}

export async function updateSchemaSpecification(
  databaseAdapter: DatabaseAdapter,
  context: TransactionContext,
  schemaSpec: AdminSchemaSpecificationUpdate
): PromiseResult<SchemaSpecificationUpdatePayload, ErrorType.BadRequest | ErrorType.Generic> {
  return await context.withTransaction(async (context) => {
    const previousSpecificationResult = await getSchemaSpecification(databaseAdapter, context);
    if (previousSpecificationResult.isError()) {
      return previousSpecificationResult;
    }
    const schema = new AdminSchema(previousSpecificationResult.value);
    const mergeResult = schema.mergeWith(schemaSpec);
    if (mergeResult.isError()) {
      return mergeResult;
    }
    const schemaSpecification = mergeResult.value;
    // TODO return with 'none' if same as previous schema
    const updateResult = await databaseAdapter.schemaUpdateSpecification(
      context,
      schemaSpecification
    );
    if (updateResult.isError()) {
      return updateResult;
    }
    return ok({ effect: 'updated', schemaSpecification });
  });
}
