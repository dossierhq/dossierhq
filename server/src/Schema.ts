import type { PromiseResult, ErrorType } from '@datadata/core';
import { ok, Schema } from '@datadata/core';
import type { Context, SessionContext } from '.';
import * as Db from './Database';
import type { SchemaVersionsTable } from './DatabaseTables';

export async function getSchema(context: Context<unknown>): Promise<Schema> {
  const { specification } = await Db.queryOne<Pick<SchemaVersionsTable, 'specification'>>(
    context,
    'SELECT specification FROM schema_versions ORDER BY id DESC LIMIT 1'
  );

  return new Schema(specification);
}

export async function setSchema(
  context: SessionContext,
  schema: Schema
): PromiseResult<void, ErrorType.BadRequest> {
  const validation = schema.validate();
  if (validation.isError()) {
    return validation;
  }
  // TODO check if different
  await Db.queryNone(context, 'INSERT INTO schema_versions (specification) VALUES ($1)', [
    schema.spec,
  ]);
  return ok(undefined);
}
