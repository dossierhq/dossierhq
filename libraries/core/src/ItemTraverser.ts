import type {
  AdminEntity,
  AdminSchema,
  AdminValueTypeSpecification,
  FieldSpecification,
  ItemValuePath,
  ValueItem,
} from '.';
import { isItemValueItem, isRichTextItemField, isValueTypeItemField, RichTextBlockType } from '.';

export enum ItemTraverseNodeType {
  error = 'error',
  field = 'field',
  valueItem = 'valueItem',
}

export type ItemTraverseNode =
  | ItemTraverseNodeError
  | ItemTraverseNodeField
  | ItemTraverseNodeValueItem;

interface ItemTraverseNodeError {
  path: ItemValuePath;
  type: ItemTraverseNodeType.error;
  message: string;
}

interface ItemTraverseNodeField {
  path: ItemValuePath;
  type: ItemTraverseNodeType.field;
  fieldSpec: FieldSpecification;
  value: unknown;
}

interface ItemTraverseNodeValueItem {
  path: ItemValuePath;
  type: ItemTraverseNodeType.valueItem;
  valueSpec: AdminValueTypeSpecification;
  valueItem: ValueItem;
}

export function* traverseItem(
  schema: AdminSchema,
  path: ItemValuePath,
  item: AdminEntity | ValueItem
): Generator<ItemTraverseNode> {
  let fieldSpecs;
  let fields;
  if (!isItemValueItem(item)) {
    const entitySpec = schema.getEntityTypeSpecification(item.info.type);
    if (!entitySpec) {
      const errorNode: ItemTraverseNodeError = {
        type: ItemTraverseNodeType.error,
        path,
        message: `Couldn't find spec for entity type ${item.info.type}`,
      };
      yield errorNode;
      return;
    }
    fields = item.fields;
    fieldSpecs = entitySpec.fields;
    path = [...path, 'fields'];
  } else {
    const valueSpec = schema.getValueTypeSpecification(item.type);
    if (!valueSpec) {
      const errorNode: ItemTraverseNodeError = {
        type: ItemTraverseNodeType.error,
        path,
        message: `Couldn't find spec for value type ${item.type}`,
      };
      yield errorNode;
      return;
    }

    const valueItemNode: ItemTraverseNodeValueItem = {
      type: ItemTraverseNodeType.valueItem,
      path,
      valueSpec,
      valueItem: item,
    };
    yield valueItemNode;

    fields = item;
    fieldSpecs = valueSpec.fields;
  }

  for (const fieldSpec of fieldSpecs) {
    const fieldPath = [...path, fieldSpec.name];
    const fieldValue = fields?.[fieldSpec.name];
    const fieldNode: ItemTraverseNodeField = {
      type: ItemTraverseNodeType.field,
      path: fieldPath,
      fieldSpec,
      value: fieldValue,
    };
    yield fieldNode;
    if (fieldSpec.list) {
      if (fieldValue === null || fieldValue === undefined) {
        continue;
      }
      if (!Array.isArray(fieldValue)) {
        const errorNode: ItemTraverseNodeError = {
          type: ItemTraverseNodeType.error,
          path,
          message: `Expected list got ${typeof fieldValue}`,
        };
        yield errorNode;
        continue;
      }

      for (let i = 0; i < fieldValue.length; i += 1) {
        const fieldItemPath = [...fieldPath, i];
        const fieldItem = fieldValue[i];
        yield* traverseItemFieldValue(schema, fieldItemPath, fieldSpec, fieldItem);
      }
    } else {
      yield* traverseItemFieldValue(schema, fieldPath, fieldSpec, fieldValue);
    }
  }
}

function* traverseItemFieldValue(
  schema: AdminSchema,
  path: ItemValuePath,
  fieldSpec: FieldSpecification,
  itemValue: unknown
) {
  if (isValueTypeItemField(fieldSpec, itemValue) && itemValue) {
    yield* traverseItem(schema, path, itemValue);
  } else if (isRichTextItemField(fieldSpec, itemValue) && itemValue) {
    for (let i = 0; i < itemValue.blocks.length; i += 1) {
      const blockPath = [...path, 'blocks', i];
      const block = itemValue.blocks[i];
      if (block.type === RichTextBlockType.valueItem && block.data) {
        yield* traverseItem(schema, [...blockPath, 'data'], block.data as ValueItem);
      }
    }
  }
}
