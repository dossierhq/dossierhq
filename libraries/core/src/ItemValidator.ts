import { assertExhaustive, assertIsDefined } from './Asserts.js';
import type { ItemTraverseNode } from './ItemTraverser.js';
import { ItemTraverseNodeType } from './ItemTraverser.js';
import type { ItemValuePath } from './ItemUtils.js';
import { isStringItemField } from './ItemUtils.js';
import type {
  AdminSchema,
  RichTextFieldSpecification,
  StringFieldSpecification,
} from './Schema.js';

export interface ValidationError {
  type: 'save' | 'publish';
  path: ItemValuePath;
  message: string;
}

export interface ValidationOptions {
  validatePublish: boolean;
}

export function validateTraverseNode(
  schema: AdminSchema,
  node: ItemTraverseNode<AdminSchema>,
  { validatePublish: _1 }: ValidationOptions
): ValidationError | null {
  const nodeType = node.type;
  switch (nodeType) {
    case ItemTraverseNodeType.field:
      break;
    case ItemTraverseNodeType.fieldItem:
      if (isStringItemField(node.fieldSpec, node.value) && node.value) {
        const stringFieldSpec = node.fieldSpec as StringFieldSpecification;
        if (stringFieldSpec.matchPattern) {
          const regexp = schema.getPatternRegExp(stringFieldSpec.matchPattern);
          assertIsDefined(regexp);
          if (!regexp.test(node.value)) {
            return {
              type: 'save',
              path: node.path,
              message: `Value does not match pattern ${stringFieldSpec.matchPattern}`,
            };
          }
        }
      }
      break;
    case ItemTraverseNodeType.error:
      return { type: 'save', path: node.path, message: node.message };
    case ItemTraverseNodeType.richTextNode: {
      const richTextFieldSpec = node.fieldSpec as RichTextFieldSpecification;
      if (richTextFieldSpec.richTextNodes && richTextFieldSpec.richTextNodes.length > 0) {
        if (!richTextFieldSpec.richTextNodes.includes(node.node.type)) {
          return {
            type: 'save',
            path: node.path,
            message: `Rich text node type ${
              node.node.type
            } is not allowed in field (supported nodes: ${richTextFieldSpec.richTextNodes.join(
              ', '
            )})`,
          };
        }
      }
      break;
    }
    case ItemTraverseNodeType.valueItem:
      break;
    default:
      assertExhaustive(nodeType);
  }
  return null;
}
