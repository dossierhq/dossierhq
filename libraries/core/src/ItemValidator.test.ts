import { describe, expect, test } from 'vitest';
import { ItemTraverseNodeErrorType, traverseEntity } from './ItemTraverser.js';
import type { ValidationError, ValidationOptions } from './ItemValidator.js';
import { validateTraverseNode } from './ItemValidator.js';
import {
  createRichTextParagraphNode,
  createRichTextRootNode,
  createRichTextTextNode,
} from './RichTextUtils.js';
import { AdminSchema, FieldType } from './Schema.js';
import type { EntityLike } from './Types.js';

const schema = AdminSchema.createAndValidate({
  entityTypes: [
    {
      name: 'RichTextsEntity',
      fields: [{ name: 'anyNodes', type: FieldType.RichText }],
    },
    {
      name: 'StringsEntity',
      fields: [
        { name: 'pattern', type: FieldType.String, matchPattern: 'foo-bar-baz' },
        { name: 'patternList', type: FieldType.String, list: true, matchPattern: 'foo-bar-baz' },
      ],
    },
  ],
  patterns: [{ name: 'foo-bar-baz', pattern: '^(foo|bar|baz)$' }],
}).valueOrThrow();

function validateEntity(entity: EntityLike, options: ValidationOptions) {
  const errors: ValidationError[] = [];
  for (const node of traverseEntity(schema, ['entity'], entity)) {
    const error = validateTraverseNode(schema, node, options);
    if (error) {
      errors.push(error);
    }
  }
  return errors;
}

describe('validateTraverseNode', () => {
  test('error', () => {
    expect(
      validateTraverseNode(
        schema,
        {
          type: 'error',
          path: ['entity', 'foo'],
          errorType: ItemTraverseNodeErrorType.generic,
          message: 'Error message',
        },
        { validatePublish: true }
      )
    ).toMatchInlineSnapshot(`
      {
        "message": "Error message",
        "path": [
          "entity",
          "foo",
        ],
        "type": "save",
      }
    `);
  });

  test('Pass: matchPattern matched string', () => {
    expect(
      validateEntity(
        { info: { type: 'StringsEntity' }, fields: { pattern: 'baz' } },
        { validatePublish: false }
      )
    ).toEqual([]);
  });

  test('Fail: matchPattern unmatched string', () => {
    expect(
      validateEntity(
        { info: { type: 'StringsEntity' }, fields: { pattern: 'will not match' } },
        { validatePublish: false }
      )
    ).toMatchInlineSnapshot(`
      [
        {
          "message": "Value does not match pattern foo-bar-baz",
          "path": [
            "entity",
            "fields",
            "pattern",
          ],
          "type": "save",
        },
      ]
    `);
  });

  test('Pass: matchPattern matched string items in list', () => {
    expect(
      validateEntity(
        {
          info: { type: 'StringsEntity' },
          fields: { patternList: ['foo', 'bar', 'baz'] },
        },
        { validatePublish: false }
      )
    ).toEqual([]);
  });

  test('Fail: matchPattern unmatched string item in list', () => {
    expect(
      validateEntity(
        {
          info: { type: 'StringsEntity' },
          fields: { patternList: ['foo', 'will not match'] },
        },
        { validatePublish: false }
      )
    ).toMatchInlineSnapshot(`
      [
        {
          "message": "Value does not match pattern foo-bar-baz",
          "path": [
            "entity",
            "fields",
            "patternList",
            1,
          ],
          "type": "save",
        },
      ]
    `);
  });

  test('Fail: rich text text node with line break', () => {
    expect(
      validateEntity(
        {
          info: { type: 'RichTextsEntity' },
          fields: {
            anyNodes: createRichTextRootNode([
              createRichTextParagraphNode([createRichTextTextNode('hello\nworld')]),
            ]),
          },
        },
        { validatePublish: false }
      )
    ).toMatchInlineSnapshot(`
      [
        {
          "message": "Rich text text nodes cannot contain line breaks, use linebreak nodes instead",
          "path": [
            "entity",
            "fields",
            "anyNodes",
            0,
            0,
          ],
          "type": "save",
        },
      ]
    `);
  });
});
