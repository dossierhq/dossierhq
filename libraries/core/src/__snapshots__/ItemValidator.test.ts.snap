// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Validate entity > Fail: admin only value item in normal field 1`] = `
[
  {
    "message": "Value item of type AdminOnlyValueItem is adminOnly",
    "path": [
      "entity",
      "fields",
      "any",
    ],
    "type": "publish",
  },
]
`;

exports[`Validate entity > Fail: matchPattern unmatched string 1`] = `
[
  {
    "message": "Value does not match pattern foo-bar-baz",
    "path": [
      "entity",
      "fields",
      "pattern",
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity > Fail: matchPattern unmatched string item in list 1`] = `
[
  {
    "message": "Value does not match pattern foo-bar-baz",
    "path": [
      "entity",
      "fields",
      "patternList",
      1,
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity > Fail: required with no value 1`] = `
[
  {
    "message": "Required field is empty",
    "path": [
      "entity",
      "fields",
      "required",
    ],
    "type": "publish",
  },
]
`;

exports[`Validate entity > Fail: rich text text node with line break 1`] = `
[
  {
    "message": "Rich text text nodes cannot contain line breaks, use linebreak nodes instead",
    "path": [
      "entity",
      "fields",
      "anyNodes",
      0,
      0,
    ],
    "type": "save",
  },
]
`;

exports[`validateTraverseNodeForSave > error 1`] = `
{
  "message": "Error message",
  "path": [
    "entity",
    "foo",
  ],
  "type": "save",
}
`;
